File: exo_stand_status/scripts/save_listener.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# 接收/joint_states/position的话题消息。10s后，自动停止接收消息，并保存到同一工作空间目录下的joint_states.txt，在读取txt文件内容，保存到四个数组中

import rospy
from sensor_msgs.msg import JointState
import time

array1 = []
array2 = []
array3 = []
array4 = []

def joint_states_callback(msg):
    global array1, array2, array3, array4
    with open("joint_states.txt", "a") as file:
        for i in range(len(msg.position)):
            file.write(str(msg.position[i]))
            if i < len(msg.position) - 1:
                file.write(", ")
        file.write("\n")

def main():
    rospy.init_node("joint_states_subscriber")
    sub = rospy.Subscriber("/joint_states", JointState, joint_states_callback)
    
    rate = rospy.Rate(10)  # 10 Hz
    
    start_time = rospy.Time.now()
    duration = rospy.Duration(10.0)  # 10 seconds
    
    while rospy.Time.now() - start_time < duration:
        rate.sleep()
    
    # Read data from file
    with open("joint_states.txt", "r") as file:
        for line in file:
            values = line.strip().split(", ")
            if len(values) == 4:
                array1.append(float(values[0]))
                array2.append(float(values[1]))
                array3.append(float(values[2]))
                array4.append(float(values[3]))
    
    # Print the arrays
    print("joint1:", " ".join(map(str, array1)))
    print("joint2:", " ".join(map(str, array2)))
    print("joint3:", " ".join(map(str, array3)))
    print("joint4:", " ".join(map(str, array4)))

if __name__ == "__main__":
    main()

File: exo_stand_status/src/save_listener.cpp
#include <ros/ros.h>
#include <sensor_msgs/JointState.h>
#include <fstream>
#include <chrono>
#include <thread>

// 接收/joint_states/position的话题消息。10s后，自动停止接收消息，并保存到同一工作空间目录下的joint_states.txt，在读取txt文件内容，保存到四个数组中

std::vector<double> array1, array2, array3, array4;

void jointStatesCallback(const sensor_msgs::JointState::ConstPtr& msg) {
    std::ofstream file("joint_states.txt", std::ios_base::app);
    if (!file.is_open()) {
        ROS_ERROR_STREAM("Could not open the file.");
        return;
    }

    for (size_t i = 0; i < msg->position.size(); ++i) {
        file << msg->position[i];
        if (i < msg->position.size() - 1)
            file << ", ";
    }
    file << std::endl;

    file.close();
}

int main(int argc, char** argv) {
    ros::init(argc, argv, "joint_states_subscriber");
    ros::NodeHandle nh;

    ros::Subscriber sub = nh.subscribe("/joint_states", 10, jointStatesCallback);

    ros::Rate rate(10); // 10 Hz

    // Wait for 10 seconds
    for (int i = 0; i < 100; ++i) {
        ros::spinOnce();
        rate.sleep();
    }

    // Read data from file
    std::ifstream file("joint_states.txt");
    if (!file.is_open()) {
        ROS_ERROR_STREAM("Could not open the file.");
        return -1;
    }

    std::string line;
    while (std::getline(file, line)) {
        double val1, val2, val3, val4;
        if (sscanf(line.c_str(), "%lf, %lf, %lf, %lf", &val1, &val2, &val3, &val4) == 4) {
            array1.push_back(val1);
            array2.push_back(val2);
            array3.push_back(val3);
            array4.push_back(val4);
        }
    }

    // Print the arrays
    std::cout << "joint1: ";
    for (double val : array1)
        std::cout << val << " ";
    std::cout << std::endl;

    std::cout << "joint2: ";
    for (double val : array2)
        std::cout << val << " ";
    std::cout << std::endl;

    std::cout << "joint3: ";
    for (double val : array3)
        std::cout << val << " ";
    std::cout << std::endl;

    std::cout << "joint4: ";
    for (double val : array4)
        std::cout << val << " ";
    std::cout << std::endl;

    return 0;
}
